python/lsst/resources/schemeless.py

```diff
@@ -70,9 +72,38 @@ def abspath(self) -> ResourcePath:
         # the options that will force the code below in _fixupPathUri to
         # return a file URI from a scheme-less one.
         return ResourcePath(
-            str(self), forceAbsolute=True, forceDirectory=self.isdir(), isTemporary=self.isTemporary
+            str(self), forceAbsolute=True, forceDirectory=self.dirLike, isTemporary=self.isTemporary
         )
 
+    def isdir(self) -> bool:
+        """Return whether this URI is a directory.
+
+        Returns
+        -------
+        isdir : `bool`
+            `True` if this URI is a directory or looks like a directory,
+            else `False`.
+
+        Notes
+        -----
+        If the URI is not known to refer to a file or a directory the file
+        system will be checked. The relative path will be resolved using
+        the current working directory. If the path can not be found, `False`
+        will be returned (matching `os.path.isdir` semantics) but the result
+        will not be stored in ``dirLike`` and will be checked again on request
+        in case the working directory has been updated.
+        """
+        if self.dirLike is None:
+            try:
+                status = os.stat(self.ospath)
+            except FileNotFoundError:
+                # Do not update dirLike flag.
+                return False
+
+            # Cache state for next time.
+            self.dirLike = stat.S_ISDIR(status.st_mode)
+        return self.dirLike
```

Comment:
This is no different to how `os.path.isdir` works in that if you give it a relative path it will look for it relative to cwd.  At issue I think is whether `dirLike is None` should ever cache the result since it becomes invalidated if you change directory.

The dirLike annotation lets you avoid relying on the cwd which, as you say, can get a bit confusing if `dir1` is a file now but if I move to a different directory it's a directory.  

Butler now does force the issue when it knows there is a file but there are some cases where the user gives us a schemeless URI (for say the location of butler root) and the thing using it doesn't know whether it's a directory or a file. Most days I will do `butler query-datasets ../pipelines_check/DATA` and butler has to work out that I meant it to use the `butler.yaml` inside `DATA` and didn't mean that I wanted it to load a file called `DATA`. The `isdir` check here with `dirLike is None` triggers in this case (and zsh has a habit of removing the trailing `/` on tab complete and so the `ResourcePath` constructor can't work it out from the trailing `/`).

Reply:
I removed the cache setting so that cwd will do the right thing if you are explicitly asking if a relative path refers to a local directory.